source/language/gates.rst

```diff
@@ -166,10 +170,57 @@ of the gate definition.
    // this is invalid:
    gate g a
    {
-     U(0, 0, 0) a[0];
+     U(0, 0, 0) a[0]; // not allowed to index an individual qubit operand
+   }
+
+'General' gate declarations have a similar structure, with minor differences:
+
+.. code-block:: c
+
+   gate name(typedParams) : typedQargs
+   {
+     body
+   }
+
+The optional parameter list ``typedParams`` is a comma-separated list of variable
+parameters, which in this case must be provided with explicit type specifications.
+The list ``typedQargs`` is a comma-separated list of operands which are either individual
+qubits, or registers/arrays of qubits, where each is again provided with an explicit type
+specification.
+If there are no variable parameters, the parentheses are optional. At least
+one quantum operand is required, and the quantum operands
+must be immediately preceded by a ``:`` delimiter.
+The arguments in ``typedQargs`` can be indexed within the body
+of the gate definition if, and only if, it is a register or array of qubits.
+
+.. code-block:: c
+
+   // this is ok:
+   gate g(angle alpha, int k): qubit[3] a
+   {
+     U(0, 0, alpha) a[0];
+     U(0, 0, alpha/k) a[1];
+     U(0, 0, alpha/(k**2)) a[2];
    }
 
-Only built-in gate statements and calls to previously defined gates can appear in ``body``.
+   // this is also ok:
+   gate qutritX : qubit[2] a {
+     x a[1];
+     cx a[1], a[0];
+     cx a[0], a[1];
+   }
+
+   // this is invalid:
+   gate g(angle alpha, int k): qubit a
+   {
+     U(0, 0, alpha) a[0]; // not allowed to index an individual qubit operand
+     U(0, 0, alpha/k) a[1];
+     U(0, 0, alpha/(k**2)) a[2];
+   }
+
+
+For either kind of gate declaration, only built-in gate statements and calls to previously
+defined gates can appear in the body of the ``gate``.
 For example, it is not valid to
 declare a classical register in a gate body. Looping constructs over these quantum
 statements are valid.
```

Comment:
This text wasn't introduced in this PR, but I just realized that these two statements are in conflict with each other because the loop induction variable is effectively declared by a `for` loop statement.